[week7] 권한 확인하기

rlawlstjd0077 opened the pull request:
# 목표
## 해야 할 일
### 로그인이 필요없는 API
- [x] 로그인 - POST /session
- [x] 회원 생성하기 - POST /user
- [x]  고양이 장난감 목록 얻기 - GET /products
- [x] 고양이 장난감 상세 조회하기 - GET /products/{id}
### 로그인이 필요한 API
- [ ] 고양이 장난감 등록하기 - POST /products
- [ ] 고양이 장난감 수정하기 - PATCH /products/{id}
- [ ] 고양이 장난감 삭제하기 - DELETE /products/{id}
- [ ] 회원 수정하기 - POST /users/{id}
- [ ] 회원 삭제하기 - DELETE /users/{id}
## 요구 사항
- [ ] 테스트 커버리지 100%를 달성해야 합니다.
- [x] 모든 API 테스트를 통과해야 합니다.
- [x] 모든 E2E 테스트를 통과해야 합니다.

트레이너님 안녕하세요 :) 이번 한주도 잘부탁드립니다 🙏

rlawlstjd0077 committed:
feat: Add spring-sequrity-starter dependency

rlawlstjd0077 committed:
feat: Remove interceptor and config

rlawlstjd0077 committed:
feat: Add authentication filter and security config

rlawlstjd0077 committed:
feat: Add authentication error filter

rlawlstjd0077 committed:
feat: Modify UNAUTHORIZED error code to be sent to authentication error filter

rlawlstjd0077 committed:
feat: Add feature generate and pass UserAuthentication in filter

rlawlstjd0077 committed:
feat: Add feature to set authentication target method as an 'PreAuthorize' annotation

rlawlstjd0077 commented:
# 2021.03.08
## 진행내역
- `AuthenticationInterceptor` -> `JwtAuthenticationFilter` 대체
## 느낀점
- 과제 영상을 여러번 봤는데도 이번주 과제는 상당히 난이도가 있어 보이네요 .. 
  - 학습할 것이 아주 많을 것 같습니다.

rlawlstjd0077 committed:
feat: Add default constuctor to dto

rlawlstjd0077 committed:
fix: Modify to use real user id

rlawlstjd0077 committed:
feat: Add 'UnAuthorizedUserAccessException' to prevent access other people's resources

rlawlstjd0077 committed:
feat: Add validate forbidden access to 'UserController'

hannut91 commented:
Spring Security가 상당히 복잡합니다 ㅎㅎㅎ 사실 외우기 보다는 이러한 것들이 있구나~ 정도로 보시고 나중에 찾아서 할 수 있을 정도면 됩니다. 
Spring Security가 Spring이랑 또 완전히 상관이 없어서 스타일이 아주 다른데, 너무 어렵게 생각하시기 보다는 사용방법만 익히는 것으로 과제를 진행하시면 좋을 것 같습니다

rlawlstjd0077 committed:
feat: Add 'EncodingFilter' to process utf-8 encoding

rlawlstjd0077 committed:
test: Modify to use 'APPLICATION_JSON' instead of 'APPLICATION_JSON_UTF_8'

rlawlstjd0077 committed:
test: Modify to set id

rlawlstjd0077 committed:
test: Apply to check forbidden user access

rlawlstjd0077 committed:
feat: Extract validate user access to PreAuthorize

rlawlstjd0077 committed:
refactor: Remove 'UnAuthorizedUserAccessException'

rlawlstjd0077 committed:
feat: Add lommbok config

rlawlstjd0077 committed:
refactor: Remove useless method

rlawlstjd0077 committed:
refactor: for test

rlawlstjd0077 committed:
fix: Remove forbidden access for not found exception

rlawlstjd0077 committed:
test: Modify test not found user

rlawlstjd0077 commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -48,7 +48,7 @@ UserResultData update(
     @ResponseStatus(HttpStatus.NO_CONTENT)
-    @PreAuthorize("isAuthenticated() and #id == authentication.principal")
+    @PreAuthorize("isAuthenticated()")
     void destroy(@PathVariable Long id) {
```
`DELETE /users/{id}` 에 대해서 본인 계정에 대해서만 삭제 되도록 구현하고자 했는데 (본인 계정이 아닌 경우 `FORBIDDEN` 응답)
API TESTCASE 에는 존재하지 않는 `ID`에 대해서 삭제했을 때 `NOT_FOUND` 응답을 하도록 기대를 하고 있어서 삭제를 했습니다.

그래서 생각해보기론 `존재하지 않는 ID 일 경우 NOT FOUND` -> `본인 계정이 아닌 경우 FORBIDDEN` 순서대로 처리를 해줘야 하나? 했는데 그러면 Controller 상에서 로직이 복잡해질것 같아 딜레마네요 .... 😂

트레이너님은 어떻게 생각하시나요 ?

rlawlstjd0077 replied:
해결했습니다 :) 

[커밋](https://github.com/CodeSoom/spring-week7-assignment-1/pull/2/commits/d05789a3274dae79901b64d3887281774abc1ea3) 에서 `@PreAuthorize` 가 아닌 `@PostAuthorize` 로 변경하면서 인증 처리 순서를 뒤로 미루는 방법으로 해당 이슈를 해결할 수 있었습니다. 

질문한게 오히려 약이 됐네요 .. ㅎㅎㅎ 감사합니다 🙂

hannut91 replied:
바로 해결을 하셨군요 ㅎㅎ API의 요구사항마다 다를 수 있을 것 같습니다. 만약에 사용자를 삭제하는 기능을 사용자가 삭제할 때는 자신만 삭제할 수 있어야 할 것 같고, 만약 관리자가 사용하는 API라면 본인 계정과 상관없이 삭제할 수 있어야 하기 때문에 `NOT FOUND`로 돌려줄 것 같습니다.

지금은 본인 계정이 아닌 경우 애초에 `FORBIDDEN`에러가 발생하기 때문에 `NOT FOUND`까지는 안가서 그런 것 같네요

rlawlstjd0077 replied:
오늘 다시 살펴보니 `@PostAuthorize` 로 변경하게 되면서 `FORBIDDEN` 체크는 할 수 있었지만 이미 삭제가 된 후에 체크를 한다는 슬픈 사실을 깨닫게 되었습니다 ..  😂
자료를 찾아보니 `@PostAuthorize' 에 대해 '사용자 삭제, 업데이트를 수행하는 메서드에는 사용하지 마십시오' 라고 하더군요.
일단 수정을 하고 말씀 해주신 대로 API 요구사항에 맞춰서 Test를 수정 하던지 로직을 수정하던지 해야 할 것 같습니다 ㅎㅎ

rlawlstjd0077 committed:
test: Add test return forbidden status when destroy with another userId

rlawlstjd0077 committed:
feat: Modify pre authorize to post authorize to destroy

rlawlstjd0077 commented:
# 2021.03.09
## 진행내역
- 로그인이 필요한 API에 `Forbidden Access` 방지 처리 추가

감사합니다 🙏

rlawlstjd0077 commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -40,16 +42,18 @@ public Product detail(@PathVariable Long id) {
 
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
+    @PreAuthorize("isAuthenticated()")
```
어노테이션 기반 Method Security 설정에서 기본으로 제공하는 표현식들을 공유합니다.
<img width="860" alt="Screen Shot 2021-03-09 at 9 19 36 PM" src="https://user-images.githubusercontent.com/18481400/110469586-2afa1800-811d-11eb-828d-6ba00dcaf6fe.png">
출처: https://docs.spring.io/spring-security/site/docs/3.0.x/reference/el-access.html#el-common-built-in

hannut91 commented:
`@PostAuthorize`와 `@PreAuthorize`는 어떤 차이가 있나요? 예를들어서 설명해주실 수 있나요?

rlawlstjd0077 commented:
> @PostAuthorize와 @PreAuthorize는 어떤 차이가 있나요? 예를들어서 설명해주실 수 있나요?

각각 예를 들어서 설명 해보았습니다. 혹시 궁금한 부분이 있으시거나 잘못된 정보가 있다면 피드백 부탁드립니다 !

### @PostAuthorize
- `@PostAuthorize` 는 메서드의 로직이 먼저 실행되도록 허용하고 실행을 끝난 이후에 어노테이션에 명시된 보안 표현식이 실행됩니다.
- `@PostAuthorize` 는 주로 DB 및 기타 소스에서 정보를 읽고 반환하는 메서드에서 쓰입니다. 
  - 업데이트, 삭제 같은 리소스에 영향을 주는 메서드에는 쓰면 안된다 !
- 예제를 보겠습니다.
  - 아래 예제는 주어진 id와 일치하는 유저를 응답하는 GET 응답 메서드입니다.
  - 제약 상황으로는 '관리자 혹은 해당 id가 본인인 경우에만 응답을 한다' 로 하겠습니다.

``` java
    @GetMapping("{id}")
    @PostAuthorize("hasRole('ADMIN') or returnObject.userId == principal.userId")
    UserResultData detail(@PathVariable Long id) {
        User user = userService.getUser(id);
        return user;
    }
```

- 메서드가 호출되면 우선 id와 일치하는 `User` 를 DB로 불러온 뒤에 응답되기 전에  `"hasRole('ADMIN') or returnObject.userId == principal.userId"` 표현식이 실행됩니다.
  - 여기서 `returnObject` 는 응답 객체를 의미합니다.
  - `@PostAuthorize`  어노테이션은 `returnObject` 라는 표현식을 사용하여 응답 객체에 접근할 수 있습니다.
- 'ADMIN' role 을 가졌거나 요청을 보낸 유저의 id와 응답 객체의 id가 일치하면 정상적으로 클라이언트에게 응답합니다.
  - 만일 그렇지 않다면 403 에러(Forbidden)로 응답합니다.

### @PreAuthorize
- `@PreAuthorize` 는 메서드의 로직이 실행되기 전에 어노테이션에 명시된 보안 표현식이 실행됩니다.
- 예제를 보겠습니다.
  - 아래 예제는 주어진 id의 유저를 주어징 유저 정보로 수정 후 응답하는 PATCH 응답 메서드입니다.
  - 제약 상황으로는 '관리자 혹은 해당 id가 본인인 경우에만 응답을 한다' 로 하겠습니다.
``` java
    @PatchMapping("{id}")
    @PretAuthorize("isAuthenticated() and (hasRole('ADMIN') or (#id == principal.id)")
    UserResultData detail(@PathVariable Long id, UserModificationData user) {
        User user = userService.updateUser(user);
        return user;
    }
```
- 메서드가 호출되면 `isAuthenticated() and (hasRole('ADMIN') or (#id == principal.id)` 표현식이 먼저 실행됩니다. 
  - `#` 은 파라미터에 접근할 수 있는 접두문장 입니다.
- 'ADMIN' role 을 가졌거나 요청을 보낸 유저의 id와 응답 객체의 id가 일치하면 메서드가 실행되고 클라이언트에게 응답합니다.
  - 만일 그렇지 않다면 403 에러(Forbidden)로 응답합니다.

rlawlstjd0077 commented:
# 2021.03.10
## 진행내역
- Spring Security 동작 방식 학습 

## 질문
### 꼭 `BasicAuthenticationFilter` 를 구현하여 써야 하는가?
- `BasicAuthenticationFilter` 의 역할이 'HTTP 요청의 Authorization 헤더를 처리하여 `SecurityContextHolder`에 처리한 결과를 담는' 역할을 한다고 알고 있습니다. 
- 그런데 자료를 찾아보니 보통 '`BasicAuthenticationFilter` 는 전달 받은 `AuthenticationManager` 를 사용하여 인증, 권한 작업을 진행한다' 라는 내용이 있어서 
- 현재 저희의 `BasicAuthenticationFilter` (`JwtAuthenticationFilter`) 는 `AuthenticationManager`를 따로 사용하지 않고 있기 때문에  꼭 `BasicAuthenticationFilter`를 구현해서 사용하지 않아도 되는 것 아닌가 ?? 하는 생각이 들었습니다 
- 이게 맞는 접근인 것인가 .. ? 하여 여쭙고 싶었습니다 !

오늘도 감사합니다 ㅎㅎ 🙏

hannut91 commented:
저도 Filter만 사용해서 구현할 수 있나 해봤는데, 잘 안되네요 ㅠ 다시 해보고 다시 말씀드리겠습니다.

rlawlstjd0077 committed:
feat: Modify 'JwtAuthenticationFilter' to extends GenericFilterBean

rlawlstjd0077 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        doAuthentication((HttpServletRequest)request, (HttpServletResponse)response, chain);
+    }
```
> 저도 Filter만 사용해서 구현할 수 있나 해봤는데, 잘 안되네요 ㅠ 다시 해보고 다시 말씀드리겠습니다.

저도 한번 시도를 해보다가 `GenericFilterBean` 으로 구현해봤습니다 ! 잘 되는 것 같네요 ㅎㅎ

hannut91 replied:
아에 직접 만들고 싶었는데, 공부해야될게 너무 많네요 ㅎㅎ

rlawlstjd0077 commented:
Spring Security Filter Chain 에 대해서 공부한 내용을 공유합니다 
혹시 궁금한 점이나 잘못된 정보가 있으면 피드백 부탁드립니다. 🙂

## Spring Security Filter Chain
- 웹 티어의 Spring Security의 보안 지원은 **Servlet Filter** 를 기반으로 합니다.
- Spring Security Filter Chain은 Servlet Filter Chain 중 하나의 Chain 이며 그 안에는 다시 여러 Security Filter 들이 Chain을 이루는 구조입니다.
    - 아래 그림을 보시면 이해가 쉬울 것입니다 ..
<img width="400" alt="Screen Shot 2021-03-10 at 4 01 57 PM" src="https://user-images.githubusercontent.com/18481400/110763755-9109ab00-8295-11eb-99b2-5cdc4697f3c2.png" style="display: block; margin: 0 auto;">

- Spring에서는 Servlet Container Lifecycle(web.xml)과 ApplicationContext 사이를 연결할 수 있는 `DelegatingFilterProxy` 라는 Filter를 제공합니다.
- 실제로는 `DelegatingFilterProxy` 가 감싸고 있는 `FilterChainProxy` 에 의해서 수행됩니다.
- `FilterChainProxy` 는 Security Filter Chain 을 가지고 이있으면서 이들을 순회하면서 실행합니다.
    - 그렇습니다. 저 `SecutiryFilterChain`  중간에는 제가 선언한 `JwtAuthenticationFilter` 도 있습니다 !

## Code를 보자. 
- 저 말만 봐서는 잘 와닿질 않습니다. (제가 썼는데 저 조차도 이해가 안됩니다 .. 😂)

- 제 `JwtAuthenticationFilter` 의 `doFilter()`메서드에 디버깅을 걸어보겠습니다.
<img width="921" alt="Screen Shot 2021-03-11 at 5 47 03 PM" src="https://user-images.githubusercontent.com/18481400/110764137-04abb800-8296-11eb-92df-adabb3993c43.png">


### FilterChainProxy
- 그런다음 파라미터 `chain` 을 살펴보겠습니다.
- 우선 `chain` 인스턴스는 `FilterChainProxy` 인스턴스임을 알 수 있습니다.
    - 저 위의 그림에 `SecurityFilterChain` 을 들고 있는 녀석이죠
- 그리고`originalChain` 은 Servlet 자체의 필터 Chain을 말하는 것 같습니다.
    - Chain 안을 들여다 보면 `DelegatingFilterProxy` 가 3번 인덱스에 위치하는 것을 볼 수 있습니다 !
    - 저 위에 `FilterChainProxy` 을 들고 있는 녀석이죠
<img width="1063" alt="Screen Shot 2021-03-11 at 5 51 00 PM" src="https://user-images.githubusercontent.com/18481400/110764265-2d33b200-8296-11eb-9566-f8ce6bd1f980.png">


- `chain` 이 `FilterProxy` 인스턴스라고 하면 Security Filter Chain을 가지고 있겠죠 ?
- `chain` 의 `additionalFilters` 필드를 보면 제가 추가한 `JwtAuthenticationFilter` 가 6번 인덱스에 위치하는 것을 볼 수 있습니다!
<img width="878" alt="Screen Shot 2021-03-11 at 5 55 24 PM" src="https://user-images.githubusercontent.com/18481400/110764305-3ae93780-8296-11eb-89c4-ab072f9a21a4.png">


### VirtualFilterChain
- 그런데 가만보면 `chain` 이 `FilterChainProxy` 의 `VirtualFilterChain` 인스턴스라고 이야기 하고 있습니다.
    - 한번 가보죠.
- `VirtualFilterChain` 은  `FilterChainProxy` 클래스의 이너 클래스로 선언되어 있습니다.
<img width="655" alt="Screen Shot 2021-03-11 at 5 58 43 PM" src="https://user-images.githubusercontent.com/18481400/110764381-56544280-8296-11eb-883d-38fd7b7863cd.png">


- `FilterChainProxy` 가 `VirtualFilterChain` 을 어떻게 사용하는지 보면 명확합니다.
- `FilterChainProxy` 자체도 Filter 이기 때문에(`GenericFilterBean` 의 구현체) `doFilter()` 가 호출됩니다.
    - `doFilter()` 는 `DelegatingFilterProxy` 에서 호출하죠
- 아래 메서드에서 중요한 부분은 `doFilterInternal()` 메서드 호출입니다.
<img width="542" alt="Screen Shot 2021-03-11 at 6 03 18 PM" src="https://user-images.githubusercontent.com/18481400/110764445-6b30d600-8296-11eb-96df-6a0d81cd2244.png">


- 메서드 하단을 보면 `VirtualFilterChain` 객체를 생성하고 `doFilter` 를 실행해줍니다.
    - 이때 filter 목록을 넣어주는 것 같네요
<img width="619" alt="Screen Shot 2021-03-11 at 6 04 53 PM" src="https://user-images.githubusercontent.com/18481400/110764551-8a2f6800-8296-11eb-9537-dd7c18bfaf68.png">

<br/>

- 이제 모든 비밀이 풀린 것 같습니다 !
- `VirtualFilterChain` 의 `doFilter()` 에서는 주입받은 Filter list (Security Filter Chain) 을 순차적으로 실행줍니다.
    - 여기서 제가 선언한 `JwtAuthenticationFilter` 가 중간에 실행됩니다
- 그리고 filter를 전부 순회한 경우 `originalChain` (Servlet Filter Chain) 으로 돌아가 다음 Filter를 호출 합니다.
<img width="713" alt="Screen Shot 2021-03-11 at 6 07 24 PM" src="https://user-images.githubusercontent.com/18481400/110764599-961b2a00-8296-11eb-899e-3f8d90c6543c.png">

rlawlstjd0077 committed:
refactor: Add javadoc to entire class

rlawlstjd0077 committed:
test: Remove destroy another user test

rlawlstjd0077 committed:
test: Add register duplicated email test

rlawlstjd0077 committed:
feat: Remove useless exception handler

rlawlstjd0077 committed:
refactor: Remove useless controller error handler

rlawlstjd0077 committed:
refactor: Modify imports

rlawlstjd0077 commented:
# 2021-03-11
## 진행내역
- 전체 클래스 JavaDoc 작성
- `JwtAuthenticationFilter` 슈퍼 클래스 `GenericFilterBean` 로 수정

감사합니다 :)

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityConfig.java:
```diff
@@ -0,0 +1,55 @@
+
+        http
+                .csrf().disable()
+                .addFilterBefore(authenticationFilter, UsernamePasswordAuthenticationFilter.class)
```
기존에는 `addFilter`였는데, order가 지정되지 않아서 `addFilterBefore`를 사용하신 것 같네요 ㅎㅎ 어떤 차이가 있을까요?

rlawlstjd0077 replied:
맞습니다 이부분도 공유를 하고 싶었던 내용이었습니다 ! ㅎㅎ

## addFilter
- `addFilter` 메서드는 [JavaDoc](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/config/annotation/web/builders/HttpSecurity.html#addFilter-javax.servlet.Filter-) 에서도 볼 수 있듯이 'Spring Security에서 제공하는 Filter 클래스, 혹은 이를 확장한 클래스의 인스턴스'만 추가할 수 있습니다!
- 'Spring Security에서 제공하는 Filter 클래스'는 요런 클래스들이 제공된다고 합니다 
<img width="383" alt="Screen Shot 2021-03-12 at 1 55 19 PM" src="https://user-images.githubusercontent.com/18481400/110894119-a389f000-833a-11eb-814c-bad423368367.png">

- 실제 코드를 보게 되면 주어진 filter가 `comparator.isRegistered(filterClass)`를 통해서 등록되어 있는 필터 인지를 확인합니다.
  - 만약 아니라면 예외를 던집니다 !
  - 사실 저도 무턱대고 커스텀 필터로 `addFilter()` 를 사용하다 예외를 만나는 바람에 자세하게 알아보게 되었습니다 ㅎㅎ
<img width="776" alt="Screen Shot 2021-03-12 at 1 59 13 PM" src="https://user-images.githubusercontent.com/18481400/110894515-58bca800-833b-11eb-98ba-69bd392a3d25.png">

## addFilterBefore
- `addFilterBefore` 메서드는 Servlet 의 `Filter` 인터페이스를 구현한 필터라면 추가할 수 있고 이때 Security Filter Chain에 이미 등록되어 있는 Filter 클래스의 앞쪽 순서로 추가가 됩니다.
  - 다만 이때도 대상이 되는 `beforeFilter` 가 Chain에 존재하지 않는 Filter 라면 예외를 던집니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityConfig.java:
```diff
@@ -0,0 +1,55 @@
+                .exceptionHandling()
+                .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
+
+        http
+                .addFilterBefore(encodingFilter, AuthenticationErrorFilter.class);
```
인코딩을 필터로 추가해주셨군요

```kotlin
@Bean
fun characterEncodingFilter() = CharacterEncodingFilter().apply {
    encoding = "UTF-8"
    setForceEncoding(true)
}
```
저는 위와 같이 설정을 잡아줬었어요. 이건 코틀린 예제이긴 하지만, 자바도 똑같이 있을 것 같아요

rlawlstjd0077 replied:
저도 추가하면서 Security Filter 사이에 있는게 조금 이상 하긴 했었는데 .. 이런 간단한 방법이 있었군요 !

rlawlstjd0077 committed:
feat: Modify to store encrypted user password
DB에 저장되는 유저 비밀번호에 대해서 평문 그대로 저장하도록 구현되어 있었습니다.
이는 비밀번호가 노출되었을 때 심각한 문제가 발생할 수 있습니다.
따라서, DB에 저장되는 유저 비밀번호를 암호화 하여 저장하도록 수정하였습니다.

rlawlstjd0077 committed:
feat: Modify encoding filter to create as bean

rlawlstjd0077 commented:
# 2021-03-12
## 진행내역
- 사용자 비밀번호 암호화 처리 추가


감사합니다 :)

rlawlstjd0077 committed:
feat: Remove encoding filter

rlawlstjd0077 committed:
refactor: Fix build error

hannut91 commented:
커밋을 보면서 더 표현력을 높일 수 있을까? 생각해보았습니다.

> feat: Modify to store encrypted user password
>
> DB에 저장되는 유저 비밀번호에 대해서 평문 그대로 저장하도록 구현되어 있었습니다.
> 이는 비밀번호가 노출되었을 때 심각한 문제가 발생할 수 있습니다.
> 따라서, DB에 저장되는 유저 비밀번호를 암호화 하여 저장하도록 수정하였습니다.

`Modify`같이 `수정`이라는 말은 너무 많은 의미를 가지고 있어서 지양하는게 좋을 것 같아요. 수정은 버그 수정, 기능 추가, 코드 개선 등등 모든 의미로 쓰일수가 있어서 `Encrypt user password`처럼 원래 하려고 했던 일을 적어줘도 좋을 것 같아요

> feat: Modify encoding filter to create as bean

`Register encoding filter as bean` 이렇게 하는 것은 어떨까요?

> feat: Remove encoding filter
> refactor: Fix build error

```
Remove unused encoding filter

인코딩 필터를 필터로 등록했었는데, Bean으로 등록해서 사용하는 것으로 변경하여 더 이상 사용되지 않아 삭제했습니다.
```

저라면 어떻게 할까 고민하다가 위와 같이 써봤습니다.

rlawlstjd0077 commented:
트레이너님 피드백 감사합니다 🙏🙏
주신 피드백을 보고 느낀점은 Commit Message Summary에는 '한 것'에 대한 메시지가 간결하면서도 어느정도 구체적인 것이 표현력을 더 좋게 하는 것 같습니다. (그 부분이 현재 제가 부족한 점 이구요 .. ㅎㅎ ) 

> Modify같이 수정이라는 말은 너무 많은 의미를 가지고 있어서 지양하는게 좋을 것 같아요. 수정은 버그 수정, 기능 추가, 코드 개선 등등 모든 의미로 쓰일수가 있어서 Encrypt user password처럼 원래 하려고 했던 일을 적어줘도 좋을 것 같아요

트레이너님 피드백을 듣고 보니 확실히 `수정` 이라는 용어에 모호함이 보이는 것 같습니다. (사실 여태껏 쓰면서도 찜찜함이 있었지만 .. ) 
앞으로는 `수정`을 비롯해서 애매한 용어들에 대해서는 조금더 명확하고 구체적인 표현이 없을까 고민해봐야 할 것 같습니다 !

> Register encoding filter as bean 이렇게 하는 것은 어떨까요?

동의 합니다 !

> Remove unused encoding filter
인코딩 필터를 필터로 등록했었는데, Bean으로 등록해서 사용하는 것으로 변경하여 더 이상 사용되지 않아 삭제했습니다.
저라면 어떻게 할까 고민하다가 위와 같이 써봤습니다.

제가 볼 때도 좀 더 구체적으로 표현해주니 표현력이 훨씬 좋아지는 것 같습니다 !

hannut91 commented:
커밋 메세지를 제대로 남기려고만 해도 이번 작업에서 내가 무엇을 하려고 하는지가 명확해져서, 커뮤니케이션 이상의 효과를 가져오는 것 같습니다 ㅎㅎ 저도 자극 받아서 커밋 메시지와 풀 리퀘스트 메시지를 제대로 작성하려는 노력을 더 해보겠습니다.
이번주 고생 많으셨습니다!
